alternator/streams.cc

```diff
@@ -182,5 +182,277 @@ future<alternator::executor::request_return_type> alternator::executor::list_str
+        type = stream_view_type::NEW_AND_OLD_IMAGES;
+    } else if (opts.preimage()) {
+        type = stream_view_type::OLD_IMAGE;
+    } else if (opts.postimage()) {
+        type = stream_view_type::NEW_IMAGE;
+    }
+    return type;
+}
+
+}
+
+template<typename ValueType>
+struct rapidjson::internal::TypeHelper<ValueType, alternator::stream_view_type>
+    : public from_string_helper<ValueType, alternator::stream_view_type>
+{};
+
+namespace alternator {
+
+using namespace std::chrono_literals;
+using namespace std::string_literals;
+
+// the time window we enforce in querying cdc log entries.
+// idea is that write timestamp variations within cluster 
+// will be within this frame, and a timestamp limit set 
+// to maximum now() - confidence_interval will get all records 
+// up until that limit, regardless of late writes.
+static constexpr auto confidence_interval = 30s;
+
+/**
+ * This implementation of CDC stream to dynamo shard mapping
+ * is simply a 1:1 mapping, id=shard. Simple, and fairly easy
+ * to do ranged queries later.
+ * 
+ * Downside is that we get a _lot_ of shards. And with actual
+ * generations, we'll get hubba-hubba loads of them. 
+ * 
+ * Improvement ideas: 
+ * 1.) Prune expired streams that cannot have records (check ttls)
+ * 2.) Group cdc streams into N per shards (by token range)
+ * 
+ * The latter however makes it impossible to do a simple 
+ * range query with limit (we need to do staggered querying, because
+ * of how GetRecords work).
+ * The results of a select from cdc log on more than one cdc stream,
+ * where timestamp < T will typically look like: 
+ * 
+ * <shard 0> <ts0> ...
+ * ....
+ * <shard 0> <tsN> ...
+ * <shard 1> <ts0p> ...
+ * ...
+ * <shard 1> <tsNp> ...
+ *
+ * Where tsN < timestamp and tsNp < timestamp, but range [ts0-tsN] and [ts0p-tsNp]
+ * overlap. 
+ * 
+ * Now, if we have limit=X, we might end on a row inside shard 0.
+ * For our next query, we would in the simple case just say "timestamp > ts3 and timestamp < T",
+ * but of course this would then miss rows in shard 1 and above. 
+ * 
+ * However if we instead set the conditions to 
+ * 
+ *   "(token(stream_id) = token(shard 0) and timestamp > ts3 and timestamp < T)
+ *      or (token(stream_id) > token(shard 0) and timestamp < T)"
+ * 
+ * we can get the correct data. Question is however how well this will perform, if at 
+ * all.
+ * 
+ * For now, go simple, but maybe consider if this latter approach would work
+ * and perform. 
+ * 
+ */
+
+future<executor::request_return_type> executor::describe_stream(client_state& client_state, service_permit permit, rjson::value request) {
+    _stats.api_operations.describe_stream++;
+
+    auto limit = rjson::get_opt<int>(request, "Limit").value_or(100); // according to spec
+    auto ret = rjson::empty_object();
+    auto stream_desc = rjson::empty_object();
+    // local java dynamodb server does _not_ validate the arn. But "real" one does. 
```

Comment:
The validation is done in the parsing. I think it is worth noting where we are compatible with one and not the other (local/real), because one (me) running the local one and comparing with alternator might think "oh, we do it wrong". The comment will hopefully allow the code reader (me) to realize that it is in fact the local dynamo that behaves borked.

Reply:
Ok. My point is that we shouldn't run the local DynamoDB and nobody else should ever run the local DynamoDB, so how it behaves doesn't matter. But I guess a comment about the local DynamoDB doesn't hurt, though, even if we never plan to use it.